cep-plugin-infrastructure.md

```diff
@@ -0,0 +1,103 @@
+<table>
+<tr><td> Title </td><td> Implement initial conda plugin mechanism </td>
+<tr><td> Status </td><td> Draft  </td></tr>
+<tr><td> Author(s) </td><td> Bianca Henderson, Jannis Leidel </td></tr>
+<tr><td> Created </td><td> July 5, 2022 </td></tr>
+<tr><td> Updated </td><td> July 6, 2022 </td></tr>
+<tr><td> Discussion </td><td>https://github.com/conda-incubator/ceps/pull/32</td></tr>
+<tr><td> Implementation </td><td>https://github.com/conda/conda/pull/11435</td></tr>
+</table>
+
+## Abstract
+
+In order to enable customization and extra features that are compatible with and discoverable by `conda` (but do not necessarily ship as a default part of the `conda` codebase), we would like to implement an official `conda` plugin mechanism.
+
+A plugin mechanism in `conda` would provide many benefits, including (but not limited to) the coverage of underserved use cases, the added ability to extend `conda` internals via official APIs, as well as lowering the barrier of entry for contributions from other stakeholders in the `conda` community and ecosystem.
+
+
+## Specification
+
+Plugins in `conda` will integrate the "hook + entry point" structure by utilizing the [`pluggy`](https://pluggy.readthedocs.io/en/stable/index.html) Python framework. This implementation can be broken down via the following two steps:
+
+- Define the hook(s) to be registered
+- Register the plugin under the `conda` entrypoint namespace
+
+
+### Hook
+
+Below is an example of a very basic plugin "hook":
+
+
+_my_plugin.py_
+```python
+import conda.plugins
+
+
+@conda.plugins.register
+def conda_subcommands():
+    ...
+```
+
+### Entrypoint namespace
+
+Below is an example of an entry point namespace for the custom plugin function, decorated with the plugin hook shown above:
+
+_setup.py_
+```python
+from setuptools import setup
+
+setup(
+    name="my-conda-plugin",
+    install_requires="conda",
+    entry_points={"conda": ["my-conda-plugin = my_plugin"]},
+    py_modules=["my_plugin"],
+)
+```
```

Comment:
Since this CEP should be looking into the future of packaging a bit, I think in this section it's actually worth adding a `pyproject.toml` example next to the backward-looking setup.py example.. and setuptools (among others) supports it now: https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html

Alternatively, we could use [flit](https://flit.pypa.io/en/latest/pyproject_toml.html#pyproject-project-entrypoints) since that's also relatively widely used in data science circles. conda-libmamba-solver [uses flit as well](https://github.com/conda-incubator/conda-libmamba-solver/blob/main/pyproject.toml). What do you think?

Reply:
I've added a separate example using a `pyproject.toml` file as suggested, please let me know what you think!

Reply:
Looks good to me!